index/segment/mem/searcher.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+)
+
+var (
+	errFirstFilterMustNotBeNegation = errors.New("first filter must be non-negation")
+	errEmptyQuery                   = errors.New("empty query specified")
+)
+
+type negationMergeFn func(x segment.PostingsList, y segment.ImmutablePostingsList)
+
+func differenceNegationFn(x segment.PostingsList, y segment.ImmutablePostingsList) { x.Difference(y) }
+func unionNegationFn(x segment.PostingsList, y segment.ImmutablePostingsList)      { x.Union(y) }
+
+type sequentialSearcher struct {
+	queryable       queryable
+	negationMergeFn negationMergeFn
+	pool            segment.PostingsListPool
+}
+
+// newSequentialSearcher returns a new sequential searcher.
+func newSequentialSearcher(
+	q queryable,
+	negationMergeFn negationMergeFn,
+	pool segment.PostingsListPool,
+) searcher {
+	return &sequentialSearcher{
+		queryable:       q,
+		negationMergeFn: negationMergeFn,
+		pool:            pool,
+	}
+}
+
+func (s *sequentialSearcher) Query(query segment.Query) (
+	candidateDocIDs segment.PostingsList,
+	pendingFilterFn matchPredicate,
+	err error,
+) {
+	if err := validateQuery(query); err != nil {
+		return nil, nil, err
+	}
+
+	// order filters to ensure the first filter has no-negation
+	filters := orderFiltersByNonNegated(query.Filters)
+	sort.Sort(filters)
```

Comment:
This is race safe yeah?  Just pondering if anyone else holds refs to query, etc and if in the future there will be parallel searches, etc.

Reply:
yep it's only used in this one path (no other references). We'll eliminate the sort if we parallelise in the future.

Reply:
done

Reply:
Ok, np - maybe worth perhaps adding a comment w.r.t to be careful about modifying it if the parallelization is done in the future.

Reply:
nah you're right, i'll make an explicit copy of the query object to avoid this issue.

Reply:
done